### PR TITLE
Fix sigterm-forwarding

### PR DESCRIPTION
**Pull request by d-sparks**
Previously, what was happening was that the SIGTERM was being forwarded
from gearcmd to its entire process group, which sent a second, uncaught
sigterm to gearcmd.

### CODE REVIEW

**Comment by d-sparks** [comment]
I updated the unit test to actually test the functionality, and also tested manually by building a local copy, connecting it to my local gearman, running jobs and killing the gearcmd PID and verifying that its child process's sigterm handling kicks in.

**Comment by johnhuangclever** [comment]
I remember adding this back in the day (right after I joined Clever) and I dug up the original ticket that caused this change to be made:
https://clever.atlassian.net/browse/SCHOOLS-22?focusedCommentId=34890&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-34890

I think we will have to change gaprov-worker's bash script to forward the sigterm properly before we merge this change in.

**Comment by d-sparks** [comment]
@johnhuangclever That's not necessary, gaprov-worker doesn't use `--pass-sigterm`.

File: gearcmd/worker.go [hunk 1]
```diff
@@ -239,29 +239,35 @@ func (conf TaskConfig) doProcess(job baseworker.Job, envVars []string, tryCount
...
-func sigtermProcess(p *os.Process) error {
-	// kill entire group of process spawned by our cmd.Process
-	pgid, err := syscall.Getpgid(p.Pid)
-	if err != nil {
-		return fmt.Errorf("unable to get pgid, error: %s", err)
+// stopProcess kills a given process. It's second argument is a grace period, if > 0, SIGTERM will
+// be sent first folloed by SIGKILL after the grace period is over. If the grace period is 0, this
+// is simply a hard SIGKILL.
+func stopProcess(p *os.Process, gracePeriod time.Duration) error {
+	lg.InfoD("killing-pid", logger.M{"pgid": p.Pid})
```
Comment (d-sparks) [inline gearcmd/worker.go]: Don't call this `pgid`, instead call it `pid`.

File: gearcmd/worker.go [hunk 2]
```diff
@@ -239,29 +239,35 @@ func (conf TaskConfig) doProcess(job baseworker.Job, envVars []string, tryCount
...
+		time.Sleep(gracePeriod)
 	}
-	lg.InfoD("killing-pgid", logger.M{"pgid": pgid})
-	// minus sign required to kill PGIDs
-	// we use SIGTERM so that the subprocess can gracefully exit
-	if err := syscall.Kill(-pgid, syscall.SIGTERM); err != nil {
-		return fmt.Errorf("unable to kill process, error: %s", err)
+	if err := syscall.Kill(p.Pid, syscall.SIGKILL); err != nil {
+		// The graceful shutdown may have completed, so we don't error in that case.
+		if gracePeriod == 0 || !strings.Contains(err.Error(), "no such process") {
```
Comment (natebrennand) [inline gearcmd/worker.go]: this can be an exact string match, right?
https://golang.org/src/syscall/zerrors_linux_amd64.go (line 1384)

**Comment by natebrennand** [comment]
LGTM after the log is fixed

File: gearcmd/worker.go (hunk 2 above)
Comment (d-sparks) [inline gearcmd/worker.go]: The error had some other info in it, presumably the pid, but also some other jargon.  Potentially its deterministic and could be exact, but this seems fine to me?

**Pull request merged**